Install command

andreypopp opened the pull request:
Todo before merging:

* [x] **Fix tests**
* [x] Take checksum from opam into account when installing packages
* [x] Decide if we should bundle `opam-installer-bin` and `subst` as part of the esy (let's not do that)
* [x] Make sure we depend on concrete commits of submodules
* [x] Handle bzipped archives from opam
* [x] ...

Another huge PR with various improvements but mainly:

* `esy install` command which is based on pnpm installer
* conversion from OPAM to `package.json`
* Better error reporting
* realpath command

# OPAM conversion

`opam-packages-converions` dir contains a fork of opam-npm by @yunxing which transforms `opam` files into `package.json` files.

Such `package.json` contains `"esy"` key with Esy specific config and also `"opam"` key with `"url"` (optional) and `"patch"` (optional) keys.

Those generated `package.json` are meant to be distributed along with Esy and are commited into the repo. I think that's a good idea while we have semi-manual way of mapping OPAM to npm.

`esy install` command when encounters `@opam/*` packages consults `opam-packages` directory for the corresponding `package.json` files and resolves packages/versions from there.

# Trying out

For testing I published Esy on npm under `@andreypopp/esy` so you can try it on ReLayout for example:

```
% npm install -g @andreypopp/esy
% git clone https://github.com/andreypopp/ReLayout.git
% cd ReLayout
% esy install
% esy build
```

andreypopp committed:
Memoize package resolution using a realpath as a key

This makes memoization work with package installers which symlink packages form
the global store to installations inside node_modules/.

andreypopp committed:
Fix esy entry point not to allow errors

andreypopp committed:
Use babel instead of flow-remove-types

It's slower but we can use async/await which is useful for promise-heavy code
which we expect (will be followed in the next commits) install command to be.

andreypopp committed:
Update flow version

andreypopp committed:
Fix flow errors in Sandbox module

andreypopp committed:
Fix typerex-build package def

andreypopp committed:
refactor: Move commands into subdirectories

That's makes it easier to consolidate command specific modules.

andreypopp committed:
Add logging for installation command

andreypopp committed:
Use published version of pnpm

We depend on fork of pnpm which has needed hooks. We expect this fork to be
merged upstream soon.

andreypopp committed:
Blacklist ocamlbuild from converted opam packages

4.02 has it in compiler dist.

andreypopp committed:
fix(install): fix logger

The key was unique across packages.

andreypopp committed:
Bump pnpm installer

andreypopp committed:
Build esy before publishing

andreypopp committed:
Fix path pointing outside of sandbox

andreypopp committed:
Fix converting opam packages from forks

andreypopp committed:
Convert more opam packages

andreypopp committed:
Compile own realpath command.

realpath isn't available on macOS so we compile own version of it on eject.

Fixes #42

andreypopp committed:
Collect and report sandbox errors

andreypopp committed:
Fixes for types

andreypopp committed:
Fix ocamlfind opam conversion

andreypopp committed:
Fix compiling realpath command

jordwalke commented:
Wow, I'll review this more thoroughly soon.


Curious, why was the ocaml compiler moved to a dev dependency?
https://github.com/andreypopp/ReLayout/commit/9c47935903079a0c632a6dcf60c5afaf46a41e97#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R22

Dev dependencies are only installed for the top level `npm` package, but it's required to build `ReLayout` even if you depend on `ReLayout` transitively.

jordwalke commented on opam-packages/0install.json:
```diff
@@ -0,0 +1,50 @@
+        "@opam/sha": "*", 
+        "substs": "esy-ocaml-test/substs", 
+        "@opam/extlib": "*", 
+        "@opam/ocurl": "*", 
```
Are the versions supposed to be getting dropped?
https://github.com/0install/0install/blob/master/opam#L20

andreypopp replied:
I don't think so. I just imported original impl of opam-npm almost as-is and it probably drops it in some places. Need to investigate.

andreypopp commented:
> Curious, why was the ocaml compiler moved to a dev dependency?

It was moved both to peer and dev dependencies so that the top level package could decide which ocaml package to use. That's needed before we get dep overrides in sandbox config.

jordwalke commented:
> It was moved both to peer and dev dependencies so that the top level package could decide which ocaml package to use. That's needed before we get dep overrides in sandbox config.


But will every overridable package have to be anticipated as such and be listed as a dev dependency? It seems hard to predict which will require overriding, and I think by default, even if a package is not overridden - it should work.

We can make it so that the overrider is the one that has to install the second version explicitly as a dependency, and then `esy` could make sure to compile the right version. The only downside is that you might end up downloading two versions of the package - but that's tiny amount of time spent compared to compilation.

andreypopp commented:
> But will every overridable package have to be anticipated as such and be listed as a dev dependency?

Right now — yes, unless we implement package overrides as part of sandbox config (we should do it!). I left ocaml that way just because I was experimenting with different branches.

jordwalke commented:
> esy install command when encounters @opam/* packages consults opam-packages directory for the corresponding package.json files and resolves packages/versions from there.

I think this is a clever/fast way to move forward with barely any maintenance cost. (Until we get on the fly package resolving).

Just to confirm, this will still work well with the purely offline workflow where we do:

```
git clone someProject
cd someProject
esy install
scp to server
#... then on server
cd someProject
esy build
```

andreypopp commented:
> Just to confirm, this will still work well with the purely offline workflow

Yes, `esy build` should keep working offline. We actually should add a test for ejection workflow.

jordwalke commented:
> Right now — yes, unless we implement package overrides as part of sandbox config (we should do it!). I left ocaml that way just because I was experimenting with different branches.

I'm not even sure how this would work. If `ReLayout` declares ocaml as a dev dependency, then `ocamlopt` shouldn't be in the PATH when building `ReLayout`, unless `ReLayout` is the top level package right? I think it works right now because `ReLayout` is the top level package, but it might cease to work when `ReLayout` is a dependency of another app, right?

jordwalke commented:
Does this current workflow require that we have only one version of package.json committed for each converted opam package?

andreypopp commented:
> Does this current workflow require that we have only one version of package.json committed for each converted opam package?

Yes, but a `package.json` insn't an actual `package.json` you see in node package usually — it has the form `{"versions": [list of package.json files per version]}` similar to how npm stores it in the registry. So yes, we can have multiple versions of the same package if that's what are you asking.

jordwalke commented:
Curious, why use a forked `pnpm` instead of a forked `yarn` etc? That might make it easier to end up getting some of this merged into Yarn, I imagine. Also, I understand that you might have good reasons to start with pnpm. I do wish that we could simply have `esy install` generate an `npm-shrinkwrap.json` or Yarn lockfile so that it's kind of agnostic, and then also would share the download cache with `npm` or `yarn`.

andreypopp commented:
> I'm not even sure how this would work. If ReLayout declares ocaml as a dev dependency, then ocamlopt shouldn't be in the PATH when building ReLayout, unless ReLayout is the top level package right? I think it works right now because ReLayout is the top level package, but it might cease to work when ReLayout is a dependency of another app, right?

Note that ReLayout has ocaml both in peer and dev dependencies — peer deps ensure that OCaml is in env and dev deps ensure a packager installs the ocaml when ReLayout is the top level package.

andreypopp commented:
> Curious, why use a forked pnpm instead of a forked yarn etc? That might make it easier to end up getting some of this merged into Yarn, I imagine. Also, I understand that you might have good reasons to start with pnpm. I do wish that we could simply have esy install generate an npm-shrinkwrap.json or Yarn lockfile so that it's kind of agnostic, and then also would share the download cache with npm or yarn.

The most important reason is that I have the knowledge of pnpm codebase. Also pnpm was working better for me — less bugs, more space efficiency, faster installs.

I like the idea of generating shrinkwrap/lockfile to be installer agnostic.

jordwalke commented:
> Note that ReLayout has ocaml both in peer and dev dependencies — peer deps ensure that OCaml is in env and dev deps ensure a packager installs the ocaml when ReLayout is the top level package.

Oh, I see.

jordwalke commented:
> The most important reason is that I have the knowledge of pnpm codebase.

That is also a valuable reason.

> Also pnpm was working better for me — less bugs, more space efficiency, faster installs.

Supposedly yarn is really fast though - I wonder if using git urls was the culprit?

jordwalke commented:
Another thought. Depending on how `pnpm` resolves dependencies, it might be more difficult to add a constraint solver on top of it. The same may also be true of `yarn`. I believe that many constraint solving libraries we can use to resolve versions would require the package metadata to be available. That's one reason why opam often has you `opam update` to update the client's knowledge - however, I really like the `npm` style workflow. Thoughts?

andreypopp commented:
> Depending on how pnpm resolves dependencies, it might be more difficult to add a constraint solver on top of it. The same may also be true of yarn. I believe that many constraint solving libraries we can use to resolve versions would require the package metadata to be available for nearly the entire universe. 

Yes, the algo should be modified to fetch all metadata for all versions of packages recursively to construct "universe".  Fortunately npm registry returns meta for all versions for a single package on a single API call. For git repos though you'd need `git ls-remote` to list all tags and then fetch package.json files for each tag.

jordwalke commented:
> For git repos though you'd need git ls-remote to list all tags and then fetch package.json files for each tag.

Assuming we wanted to add support for constraint solving on top of tags, right? Which would be quite amazing really. A fully decentralized package manager.

andreypopp commented:
> Assuming we wanted to add support for constraint solving on top of tags, right? Which would be quite amazing really. A fully decentralized package manager.

Yes, actually I have a patch for pnpm which does that (w/o constraint solver though).

jordwalke commented on .bin/esy:
```diff
@@ -50,7 +50,10 @@ if [ "$1" == "build" ] || [ "$1" == "shell" ] || [ "$1" == "clean" ]; then
+  builtInEject
+
+elif [ "$1" == "install" ]; then
+	/usr/bin/env node $SCRIPTDIR/esy.js install $@
```
I have no clue but how would something like this work on msys? (the `/usr/bin/env` in particular). Is this being careful because the environment is scrubbed? I didn't think it had to be scrubbed until actually running the make files.

andreypopp replied:
Yes, this is not needed. Fixed.

jordwalke commented on .bin/esy.js:
```diff
@@ -1,170 +1,13 @@
-  let transformedSource = removeTypes(source)
-  super_compile.call(this, transformedSource, filename);
-};
+const DEV = process.env.ESY_DEV || !fs.existsSync(path.join(__dirname, '..', 'lib'));
```
What's the reasoning behind `__dirname/../lib` implying EASY_DEV?

andreypopp replied:
If `__dirname/../lib` is not availabe we use code from `src/`, another way to force `src/` using `ESY_DEV` env var.

jordwalke commented on .gitmodules:
```diff
@@ -0,0 +1,11 @@
+[submodule "opam-packages-conversion/forks/ocamlfind"]
+	path = opam-packages-conversion/forks/ocamlfind
+	url = https://github.com/esy-ocaml/ocamlfind.git
+	branch = esy/1.7.1
```
Is it possible to put a hash here, and then add a comment explaining what the human readable branch is?

andreypopp replied:
Ok

andreypopp committed:
Add "esy add" command

This command adds a new package to package.json and installs it.

andreypopp committed:
Fix build-eject command function signature

andreypopp committed:
Add lwt 2.7.0

jordwalke commented on opam-packages-conversion/bin/config.py:
```diff
@@ -0,0 +1,182 @@
+        'build': 'true',
+    },
+    'ocamlfind': {
+        'build': [
```
Since ocamlfind has an override here, which takes precedence? This override, or the custom `_build` script in the fork?

https://github.com/esy-ocaml/ocamlfind/commit/38052519804fcb485e21bde0aca6b89ac9fab4d1#diff-876c319af3a4cb0cb8b20fa1913434c6R1

If the `_build` script takes precedence, do we get to remove this override?

andreypopp replied:
You are right, this should be removed as this is used for opam-package.json conversion but fork already contains converted `package.json` file.

andreypopp replied:
Fixed.

jordwalke reviewed (commented):
Still reviewing, but looking good!

andreypopp committed:
opam conversion: use packages from esy-ocaml gh org

andreypopp committed:
opam conversion: drop ocamlfind override

It is not used because we maintain a fork of ocamfind

jordwalke commented on src/PackageEnvironment.js:
```diff
@@ -102,7 +102,7 @@ function relativeToSandbox(realFromPath, toPath) {
     : toPath;
   let ret = path.relative(realFromPath, toPathToUse);
-  return (ret == '0') ? "$esy__sandbox" : path.join("$esy__sandbox", ret);
+  return (ret == '0') ? "$esy__sandbox" : `$esy__sandbox/${ret}`;
```
Isn't `path.join` better since it will be more compatible with windows (uses the correct slash)? I know we shouldn't *worry* about windows, but perhaps the more closer we are the better? Or does an msys approach make it better to use `/`?

andreypopp replied:
Not sure about win + msys in this case but `path.join` produced an invalid path for packages outside of sandboxes.

jordwalke commented on package.json:
```diff
@@ -9,12 +9,25 @@
   "bin": {
     "esy": ".bin/esy"
   },
+  "files": [
```
I believe you should bump the version to `0.1.0` for `esy` right?

jordwalke commented on src/Sandbox.js:
```diff
@@ -330,12 +367,12 @@ function packageInfoKey(env: Environment, packageInfo: PackageInfo) {
     });
     if (process.env.ESY__TEST) {
-      packageInfo.__key = `${normalizedName}-${version || '0.0.0'}`;
+      packageInfo.__cachedPackageHash = `${normalizedName}-${version || '0.0.0'}`;
```
Will the env already be fully taken into account in the `hash({})` above which accepts `env`?

andreypopp replied:
Yes, `env` is used in `hash({...})` call if that what are you asking...

jordwalke replied:
So, why the need to test `if (process.env.ESY__TEST)` to determine the package hash. I thought it would already be taken into account.

andreypopp replied:
Ah, that's made for testing purposes — if `ESY__TEST` is set then esy generates dummy keys — no hash of env/deps included. That's made so we can test output with jest snapshots on different machines. We should probably just add different branches for different platforms though and test with real hashes.

jordwalke replied:
Ah, makes total sense.

jordwalke reviewed (commented):
a few more comments. Almost done reviewing!

jordwalke commented on src/installCommand/index.js:
```diff
@@ -0,0 +1,219 @@
+}
+
+async function applyPatch(packageJson: PackageJson, target: string) {
+  if (packageJson.opam.patch) {
```
Cool feature.

jordwalke commented on opam-packages-conversion/bin/lib.py:
```diff
@@ -234,8 +234,8 @@ def opamRangeToNpmRange(range):
-        "substs": "%s/substs" % (config.GH_ORG_NAME,),
-        "opam-installer-bin": "andreypopp/opam-installer-bin",
+        "substs": "esy-ocaml/substs",
+        "opam-installer-bin": "esy-ocaml/opam-installer-bin",
```
Is there a reason why we can't use https://github.com/yunxing/opam-installer-bin
?
There's so many orgs/forks I want to try to minimize duplication and start consolidating where possible.

andreypopp replied:
We could it just needs a patch — https://github.com/yunxing/opam-installer-bin/compare/master...esy-ocaml:master —

andreypopp committed:
opam conversion: blacklist xen related packages

andreypopp committed:
build-eject: fix replace-string to respect permissions

andreypopp committed:
build-eject: add TODO item

andreypopp committed:
Rewrite Sandbox to be async, handle circular deps

Rewrite Sandbox building to be async and add a check for circular deps.

Right now we fail on circular deps.

andreypopp committed:
install: fetch/unpack using wget and tar commands

This is better as Node's modules are buggy: tar-stream wasn't been able to untar
some of the opam archives.

andreypopp committed:
opam conversion: ban js_of_ocaml from depopts

andreypopp committed:
opam conversion: put files into package.json

andreypopp committed:
Add mkdirp-then missing dep

andreypopp committed:
build-eject: expose package's own path in findlib config

This is because some packages can query themselves during its own build
(cstruct for example).

andreypopp committed:
opam conversion: fix var name interpolation

For vars with a plus (+ we skip a part after the plus.

andreypopp committed:
opam conversion: expose CAML_LD_LIBRARY_PATH for more packages

andreypopp committed:
install: rm unused code

andreypopp committed:
opam conversion: more fixes to CAML_LD_LIBRARY_PATH

JaneStreet libs put shared libs to stublibs while others into installation
location.

andreypopp committed:
opam conversion: more fixes to CAML_LD_LIBRARY_PATH

andreypopp committed:
install: lower wget timeout

andreypopp committed:
opam conversion: fixes for mtime & nocrypto

andreypopp committed:
opam conversion: include checksums in package.json

andreypopp committed:
install: check checksum when downloading packages from opam

andreypopp committed:
Fix submodules to point to concrete commits

andreypopp commented:
I decided not to commit generated `package.json` (that was silly). Will try to do a proper rebase so we don't see them in history too.

andreypopp committed:
Some opam packages are still failing to convert

andreypopp committed:
Fix realpath executable exit code

Apparently some systems fail if this is not specified explicitly.

andreypopp commented:
I addressed all TODO issues I believe. @jordwalke is there anything else should be done before we pull this in?

jordwalke commented on opam-packages-conversion/bin/config.py:
```diff
@@ -28,6 +28,7 @@
     "conf-libev",
     "lablgtk",
     "ssl",
+    "mirage-xen",
```
Just curious, why do we need to blacklist xen related packages?

andreypopp replied:
Some of its deps don't convert right now (iirc it fails on assert — `|` in deps).

jordwalke commented on src/buildEjectCommand/index.js:
```diff
@@ -382,6 +382,9 @@ function buildEjectCommand(
 
       filestat = os.stat(filename)
 
+      # TODO: we probably should handle symlinks too in a special way,
```
Have we even tested symlinks yet with `esy`?

andreypopp replied:
Not yet. Though with pnpm node_modules is a symphony no farm and things seems to work. We need to a proper test case though.

jordwalke commented on src/installCommand/index.js:
```diff
@@ -132,6 +145,33 @@ async function lookupPackageCollection(packageName: string): Promise<PackageJson
+    const stage  = tempfile(filename);
+    await mkdirp(stage);
+    await mkdirp(target);
+    await execa('wget', [resolution.tarball], {cwd: stage});
```
Will everyone have `wget`? I have it installed via brew for some reason, so that makes me thing it's not installed by default on OS X or some variants of Linux.

andreypopp replied:
Good point. I was using JS code for that but `tar-stream` fails to extract some of the tarballs downloaded via opam. Probably we could use curl then?

andreypopp replied:
Just occurred to me I should have kept fetching code in js and only untar via external command. Will fix.

jordwalke commented on opam-packages-conversion/bin/config.py:
```diff
@@ -34,7 +34,8 @@
     "deriving",
     "ocamlbuild",
-    }
+    "js_of_ocaml",
```
what's the reasoning here?

andreypopp replied:
Was hitting circular deps — I think we need to implement build deps to solve that.

jordwalke replied:
Weird I would not have expected OPAM to handle circular deps either.

jordwalke commented:
This PR is amazing. Let's ship it, and fix the wget next?

andreypopp committed:
install: fetch tarballs without wget

wget isn't available out of the box on some systems

andreypopp committed:
install: add new command

andreypopp commented:
Rebased to remove generated `package.json` files from history. Waiting for CI to confirm and will merge.
